File: 347.top-k-frequent-elements.cpp
/*
 * @lc app=leetcode id=347 lang=cpp
 *
 * [347] Top K Frequent Elements
 */
#include <unordered_map>
#include <vector>

using std::unordered_map;
using std::vector;

// @lc code=start
class Solution {
public:
    // Bucket sort
    // T: O(n)
    // S: O(n)
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> f;
        const int n = nums.size();
        for (int i = 0; i < n; ++i) {
            ++f[nums[i]];
        }

        vector<vector<int>> buckets(n + 1);  // upper bound of frequency is n
        for (const auto& e : f) {
            // buckets[freq].push_back(x)
            buckets[e.second].push_back(e.first);
        }

        vector<int> res;
        res.reserve(k);
        // upper bound of frequency is n
        for (int i = n; i >= 0; --i) {
            // if (!buckets[i].empty()) {
            //     res.insert(res.end(), buckets[i].begin(), buckets[i].end());
            // }
            // if (res.size() >= k) {
            //     break;
            // }
            for (int& x : buckets[i]) {
                res.push_back(std::move(x));
                if (res.size() == k) {
                    return res;
                }
            }
        }
        // The problem specs state that "it is guaranteed that the answer is unique",
        // so no need to take the subarray from 0th element to kth element
        return res;
    }
};
// @lc code=end

File: 347.top-k-frequent-elements.java
/*
 * @lc app=leetcode id=347 lang=java
 *
 * [347] Top K Frequent Elements
 */
import java.util.*;

// @lc code=start
class Solution {
    // T: O(n)
    // S: O(n)
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freqs = new HashMap<>();
        int max_freq = 1;
        for (int x : nums) {
            int freq = freqs.getOrDefault(x, 0) + 1;
            freqs.put(x, freq);
            max_freq = Math.max(max_freq, freq);
        }
        
        // max_freq <= n
        List<Integer>[] buckets = new List[max_freq + 1];
        for (Map.Entry<Integer, Integer> e : freqs.entrySet()) {
            int freq = e.getValue();
            if (buckets[freq] == null) {
                buckets[freq] = new ArrayList<>();
            }
            buckets[freq].add(e.getKey());
        }

        int[] res = new int[k];
        int p = 0;
        for (int i = max_freq; i >= 0; --i) {
            List<Integer> sameFreqNums = buckets[i];
            if (sameFreqNums != null) {
                for (int x : sameFreqNums) {
                    res[p++] = x;
                    if (p == k) {
                        return res;
                    }
                }
            }
        }
        return res;
    }
}
// @lc code=end

